### ISSUE TITLE
Improve the release process to better handle OTP

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#3499]
I was able to release `9.4.0` following @ntwb's [instructions](https://github.com/stylelint/stylelint/issues/3456#issuecomment-407595692).

It was a bit clunky because:

1. I had to, as per the instructions, manually delete and re-checkout the `.npmrc` file
2. I had to run `NPM_CONFIG_OTP=123456 npm run release` to trigger the prerelease checks, wait for publish step to fail because the OTP would expire, then run `NPM_CONFIG_OTP=987654 npm run release -- --skip-test --skip-cleanup` to publish before the new OTP expired.

Hopefully number 1 will be addressed by @hudochenkov's [PR to npmpub](https://github.com/stylelint/stylelint/issues/3456#issuecomment-407722043).

Maybe number 2 can also be addressed in npmpub, maybe by an interactive prompt that asks you for a OTP just before the publish step?

It'd be great to keep using npmpub for [these reasons](https://github.com/stylelint/stylelint/issues/3456#issuecomment-407672298).

Thoughts?

---

_Background: the stylelint npm package maintainers have turned on two-factor auth following [ESLint's postmortem recommendations](https://eslint.org/blog/2018/07/postmortem-for-malicious-package-publishes). So we now need to supply a one-time-password when publishing npm packages._

**Comment by ntwb** [comment]
Thanks for persisting with what was for sure a frustrating process to release stylelint 9.4 @jeddy3 🥇 

npm are looking to add "time-limited tokens", it's an npm token generated that only lasts 5-10 minutes and generated using a users existing 2FA auth.

Details https://npm.community/t/time-limited-tokens/502/4

ps. Kudos to the npm team as they jumped on and proposed this idea pretty quickly after the ESLint security incident, I'm hoping this will arrive in npm soon 🤞

**Comment by jeddy3** [comment]
> npm are looking to add "time-limited tokens"

That's good to know. The current process isn't too painful. So, I think we can wait to see if these time-limited tokens land before looking for an alternative solution.

**Comment by ntwb** [comment]
|  | [npmpub](https://www.npmjs.com/package/npmpub) | [unleash](https://www.npmjs.com/package/unleash) | [np](https://www.npmjs.com/package/np) |
| --- | --- | --- | --- |
|Checks you are on the `master` branch:|  ? |? |Yes|
|Checks the working directory is "clean":| Yes| ?| Yes|
|Checks there are no unpulled changes: |No |? |Yes|
|Cleans node_modules folder and reinstalls test with:| Yes| ?| Yes|
|Runs the test suite: |Yes |Yes |Yes|
|Bumps the package.json version:| No| Yes |Yes|
|Creates the Git tag:| Yes| ? |Yes|
|Publishes to npmjs.com:| Yes |Yes| Yes|
|Pushes commits and tags to GitHub: |Yes| Yes| Yes|
|2FA Support: |No |? |Yes|
|Publishes release info from changelog:| Yes |See note 2 |See note 1|
|Supports running in "dry-run" mode: |Yes |Yes |Yes|

Notes:

1. Opens a prefilled GitHub Releases draft after publish
2. Updates your changelog according to the conventional changelog spec

Other positive features of the packages:

Unleash:
• Pro: Automatically recovers from errors by safely resetting to the git state prior to running unleash
• Pro: Allows you to preview what files will and won't be published to NPM (`--list-publishables`)

npmpub:
• Pro: Edits the tag as a GitHub release based on the new version and corresponding changelog version (using github-release-from-changelog).

np:
• Pro: Supports interactive mode, prompting for publishing options such as minor, patch, major version bump

... I'll follow up witha summary and some rudimentary testing of Unleas and np in follow up comments

**Comment by ntwb** [comment]
Unleash `./node_modules//.bin/unleash --major --dry-run`:
```shell
[19:33:00] === UNLEASH ===
[19:33:00] Utilizing dry run mode. This is a dry run of the following actions:
[19:33:00] * Incrementing to the next "major" semantic version, "14.0.0"
[19:33:00] * Creating a changelog entry for version 14.0.0 with links to the commits on github
[19:33:00] * Adding commit "docs(CHANGELOG): Update changelog"
[19:33:00] * Bumping to a major version with the commit "chore(release): release 14.0.0"
[19:33:00] * Tagging the commit with "v14.0.0"
[19:33:00] * Publishing to NPM
[19:33:00] * Pushing to origin branch master
```


Unleash `./node_modules//.bin/--major --no-push --no-publish`:

```shell
[19:36:50] === UNLEASH ===
[19:36:50] Utilizing next version for changelog:  14.0.0
[19:36:50] Using repo type:  github
[19:36:50] Generating changelog from 13.1.0 to HEAD...
[19:36:50] Parsed 0 commits.
[19:36:50]
[19:36:51]
[19:36:52] Utilizing bumpType major
[19:36:52] Retrieving latest git commit...
[19:36:52] Getting next version of type:  major
[19:36:52] Incrementing from version:  13.1.0
[19:36:52] ~~~ WARNING ~~~
[19:36:52] You've turned both publishing & pushing off. Your changelog & package files will be updated + committed locally but nothing outside your machine will change.
[19:36:52] unleash --major --no-push --no-publish
[19:36:52] Bumped 'package.json' version to: 14.0.0
[19:36:53] [master eabb2e6] chore(release): release 14.0.0
 1 file changed, 4 insertions(+), 2 deletions(-)

[19:36:53] Tagging as: v14.0.0
[19:36:53]
```

Here's the changelog changes, its empty which is unexpected
```diff
diff --git CHANGELOG.md CHANGELOG.md
index e33f8e4..3286b5c 100644
--- CHANGELOG.md
+++ CHANGELOG.md
@@ -1,3 +1,7 @@
+<a name="14.0.0"></a>
+## 14.0.0 (2019-04-18)
+
+
 # 13.1.0
 
 -   Added: Added SCSS _shared config_ `extends` tests.
```

Unleash:

After attempting to publish from a branch that was not `master` Unleash crashed as the tag had already being created in `master` on a previous test run, as such Unleash's claim it recovers from errors by restoring the Git state did not occur in this instance, expect this is a bug or unknown use case, either way it failed to detect the branch wasn't `master` and didn't recover from it.

```shell
[19:55:34] Tagging as: v14.0.0
{ Error: Command failed: git tag  -a  -m 'Tagging as v14.0.0'   'v14.0.0'
fatal: tag 'v14.0.0' already exists

    at ChildProcess.exithandler (child_process.js:294:12)
    at ChildProcess.emit (events.js:182:13)
    at ChildProcess.EventEmitter.emit (domain.js:459:23)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
  killed: false,
  code: 128,
  signal: null,
  cmd: 'git tag  -a  -m \'Tagging as v14.0.0\'   \'v14.0.0\'' }
```

Unleash appears to rely on [conventional commits](https://www.conventionalcommits.org) a little too much for use with stylelint IMHO, if stylelint were using _conventional commits_ already then quite possibly it would be a very good fit, but for now, educating all _contributors_ to adhere to a semantic versioning commit message practice would be required to be in place and operating smoothly before switching to Unleash.

**Comment by ntwb** [comment]
np: `./node_modules//.bin/np --no-yarn --no-publish`

```shell
Publish a new version of stylelint-config-wordpress (current: 13.1.0)

Commits:
- chore(package): update stylelint to version 10.0.1 (#215)  a78332e
- Update stylelint-find-rules to the latest version 🚀 (#213)  057f539
- chore(package): update eslint-plugin-jest to version 22.0.0 (#212)  8c74c5e
- chore(package): update remark-cli to version 6.0.0 (#211)  9e7ea46
- chore(package): update husky to version 1.1.2 (#210)  bbff6f1

Commit Range:
13.1.0...master

? Select semver increment or specify new version major  14.0.0
? Will bump from 13.1.0 to 14.0.0. Continue? Yes
  ❯ Git
    ✖ Check current branch
      → Not on `master` branch. Use --any-branch to publish anyway.
      Check local working tree
      Check remote history
    Cleanup
    Installing dependencies using npm
    Running tests using npm
    Bumping version using npm
    Pushing tags
    Creating release draft on GitHub

✖ Not on `master` branch. Use --any-branch to publish anyway.
```

This is great, it detected it was not on `master` and aborted.

A few other tests also went well, where it detected the repo was _dirty_, and when I had WiFi off it couldn't install dependencies and thus failed


-----
A complete dry-run test is not available in `np` as there is no option to skip the creation and push of the _tags_ to GitHub.

**Comment by ntwb** [comment]
I think `np` is looking like a good replacement for `npmpub`

We would be able to use the `np` `--no-release-draft` flag to skip generating a draft release and continue to use `github-release-from-changelog` for publishing the changelog as we currently do, just separately without `npmpub`

I've a bit more testing to do, but out of time for today

**Comment by jeddy3** [comment]
> I've a bit more testing to do, but out of time for today

Excellent stuff. This is all very interesting.

I'd love to see how [np's OTP](https://github.com/sindresorhus/np#why) works in practice. If it's an interactive prompt that works with our long-running tests then that would be superb.

**Comment by hudochenkov** [comment]
I tried `np` last two releases of stylelint-order and was quite happy with it. It's much better than `npmpub` experience I had this weekend :)

> I'd love to see how np's OTP works in practice. If it's an interactive prompt that works with our long-running tests then that would be superb.

Yes. In one of steps it asks for OTP in prompt.

For `np` we would need to add [this line](https://github.com/hudochenkov/stylelint-order/blob/b4f75511ee59c67ce0f50f1bc22c12f98301cefd/.npmrc#L2) to `.npmrc`, otherwise git tag would be `v10.1.0` (notice the “v”):

```
tag-version-prefix = ""
```
